File: solutions/go/strain/1/strain.go
package strain

type (
	Ints    []int
	Lists   [][]int
	Strings []string
)

func keep[T any](i []T, filter func(T) bool) []T {
	var result []T
	for _, v := range i {
		if filter(v) {
			result = append(result, v)
		}
	}
	return result
}

func (i Ints) Keep(filter func(int) bool) Ints {
	return keep(i, filter)
}

func (i Ints) Discard(filter func(int) bool) Ints {
	return keep(i, func(n int) bool { return !filter(n) })
}

func (l Lists) Keep(filter func([]int) bool) Lists {
	return keep(l, filter)
}

func (s Strings) Keep(filter func(string) bool) Strings {
	return keep(s, filter)
}

File: solutions/lua/strain/1/strain.lua
local function keep(xs, pred)
  local result = setmetatable({}, { __index = table })
  for _, x in ipairs(xs) do
    if pred(x) then result:insert(x) end
  end
  return result
end

local function discard(xs, pred)
  return keep(xs, function(...) return not (pred(...)) end)
end

return { keep = keep, discard = discard }
